### PR TITLE
Wrap VersionPropertiesLoader in a BuildService to decouple build logic projects

### PR DESCRIPTION
**Pull request by breskeby**
By using a BuildService we can decouple the configuration of projects but keep making
expensive operations (e.g. parsing properties file or reading minimumJavaCompiler from file) only once
per build.

This bring us closer to decouple projects and get ultimatively configuration cache compliant. In
general we will bring in more BuildServices for the main build to follow along that Pattern and
ultimatevely remove usages of allprojects and subprojects in our build.

### CODE REVIEW

**Review by pugnascotia — APPROVED** [review]

File: build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesBuildService.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            throw new GradleException("Cannot resolve minimum compiler version via VersionPropertiesBuildService", e);
+        }
+        return minimumJavaVersion;
+    }
+
+    @Override
+    public void close() throws Exception {
+    }
+
+    public interface Params extends BuildServiceParameters {
```
Comment (mark-vieira) [inline build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesBuildService.java:59]: Can we move this to the bottom of the class definition?

File: build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesPlugin.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    private BuildLayout buildLayout;
+
+    @Inject
+    VersionPropertiesPlugin(BuildLayout buildLayout) {
+        this.buildLayout = buildLayout;
+    }
+
+    @Override
+    public void apply(Project project) {
```
Comment (mark-vieira) [inline build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesPlugin.java:29]: Should we make these one of those plugins that explodes if you try to apply to anything but the root project?

File: build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesPlugin.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        // Register the service if not done yet
+        File infoPath = new File(buildLayout.getRootDirectory(), "build-tools-internal");
+        Provider<VersionPropertiesBuildService> serviceProvider = project.getGradle().getSharedServices()
+                .registerIfAbsent("versions", VersionPropertiesBuildService.class, spec -> {
+            spec.getParameters().getInfoPath().set(infoPath);
+        });
+        project.getExtensions().add("versions", serviceProvider.forUseAtConfigurationTime().get().getProperties());
```
Comment (mark-vieira) [inline build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/VersionPropertiesPlugin.java:36]: I definitely like this pattern of adding the service itself as a project extension :+1:

**Review by mark-vieira — APPROVED** [review]
LGTM

**Pull request merged**